x-pack/filebeat/input/internal/dpop/proof.go

```diff
@@ -0,0 +1,218 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package dpop
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/rsa"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"math/big"
+	"strings"
+
+	"github.com/golang-jwt/jwt/v5"
+)
+
+// ProofGenerator builds DPoP proofs for requests.
+// It supports ECDSA, RSA and ED25519 private keys.
+type ProofGenerator struct {
+	claim   Claimer
+	key     crypto.Signer
+	jwk     any
+	signing jwt.SigningMethod
+}
+
+// Claimer wraps the Claims method.
+type Claimer interface {
+	// Claims returns a jwt.RegisteredClaims that will be added
+	// to the DPoP claims held by a ProofClaims constructed by
+	// ProofGenerator.BuildProof.
+	//
+	// When json/v2 is available this will be changed to return jwt.Claims.
+	Claims() *jwt.RegisteredClaims
+}
+
+// ClaimerFunc is an adapter to allow the use of ordinary functions as a Claimer.
+type ClaimerFunc func() *jwt.RegisteredClaims
+
+func (c ClaimerFunc) Claims() *jwt.RegisteredClaims { return c() }
+
+// NewProofGenerator creates a new ProofGenerator.
+func NewProofGenerator(claim Claimer, key crypto.Signer, signing jwt.SigningMethod) (*ProofGenerator, error) {
+	if claim == nil {
+		return nil, errors.New("nil claimer")
+	}
+	if key == nil {
+		return nil, errors.New("nil private key")
+	}
+	if signing == nil {
+		return nil, errors.New("nil signing method")
+	}
+	jwk, err := buildJWKAndAlg(key)
+	if err != nil {
+		return nil, err
+	}
+	return &ProofGenerator{claim: claim, key: key, jwk: jwk, signing: signing}, nil
+}
+
+// buildJWKAndAlg constructs a JWK (public key only). Supported keys:
+//   - *ecdsa.PrivateKey
+//   - *prsa.PrivateKey
+//   - *ed25519.PrivateKey
+func buildJWKAndAlg(privateKey any) (any, error) {
+	switch k := privateKey.(type) {
+	case *ecdsa.PrivateKey:
+		return ecPublicJWK(&k.PublicKey)
+	case *rsa.PrivateKey:
+		return rsaPublicJWK(&k.PublicKey)
+	case *ed25519.PrivateKey:
+		return edPublicJWK(k.Public().(ed25519.PublicKey)) //nolint:errcheck // errcheck is wrong to call this out; if this type assertion ever failed we should probably throw, not just panic.
```

Comment:
I think you mean:
```suggestion
		return edPublicJWK(k.Public().(ed25519.PublicKey)) //nolint:errcheck // errcheck is wrong to call this out; if this type assertion ever failed we should probably panic, not just return an error.
```

Reply:
No, what I have is what I mean.

Reply:
The comment is a contrafactual referring to [this function](https://github.com/golang/go/blob/ad3ccd92e4c2ddfc1499a5e038eb0aa0012c0dda/src/runtime/panic.go#L1195-L1226) that we should not be using; it is pointing out that if we were in the state that a `*ed25519.PrivateKey` returned a value from `Public()` that was not a `ed25519.PublicKey`, something terrible has gone wrong, not limited to issues that are explained under the computing model that the physical layer is perfect. In that case we should crash in a way that cannot be recovered. This is a `throw`, necessitated by the fact that there is unreasonable use of `recover` in the beats code that attempt to cover up these kinds of outcomes.

Reply:
Found another update in https://github.com/elastic/beats/pull/47441/commits/84668ec1ac1fc8968620b29b4711af145729114a

Now this is resolved.